File: src/movegen/queen.rs
use crate::{
    bitboard::{clear_bit, get_bit},
    enums::{Color, PieceAndColor},
    position::Position,
};

use super::{BishopMoveGen, Move, RookMoveGen};

pub struct QueenMoveGen {
    rook: RookMoveGen,
    bishop: BishopMoveGen,
}

impl QueenMoveGen {
    pub fn new(rook: RookMoveGen, bishop: BishopMoveGen) -> Self {
        Self { rook, bishop }
    }

    pub fn get_attacks(&self, square: u8, occupancy: u64) -> u64 {
        self.rook.get_attacks(square, occupancy) | self.bishop.get_attacks(square, occupancy)
    }

    pub fn generate_moves(&self, position: &Position) -> Vec<Move> {
        let color = position.color_to_move();

        let piece = match color {
            Color::White => PieceAndColor::WhiteQueen,
            Color::Black => PieceAndColor::BlackQueen,
            _ => unreachable!(),
        };

        let mut moves = Vec::new();

        let mut bitboard = position.bitboards().piece(piece);

        while bitboard != 0 {
            let source_square = bitboard.trailing_zeros() as u8;

            let mut attacks = self
                .get_attacks(source_square, position.bitboards().color(Color::Both))
                & !position.bitboards().color(color);

            while attacks != 0 {
                let target_square = attacks.trailing_zeros() as u8;

                if get_bit(position.bitboards().color(Color::Both), target_square) {
                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        None,
                        true,
                        false,
                        false,
                        false,
                    ));
                } else {
                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        None,
                        false,
                        false,
                        false,
                        false,
                    ));
                }

                clear_bit(&mut attacks, target_square);
            }

            clear_bit(&mut bitboard, source_square);
        }

        moves
    }
}

File: internal/movegen/queen.go
package movegen

import (
	"github.com/notmalte/psce/internal/bitboard"
	"github.com/notmalte/psce/internal/constants"
	"github.com/notmalte/psce/internal/position"
	"math/bits"
)

type QueenMoveGen struct {
	rookMoveGen   *RookMoveGen
	bishopMoveGen *BishopMoveGen
}

func (qmg *QueenMoveGen) GetAttacks(square uint8, occupancy uint64) uint64 {
	return qmg.rookMoveGen.GetAttacks(square, occupancy) | qmg.bishopMoveGen.GetAttacks(square, occupancy)
}

func NewQueenMoveGen(rmg *RookMoveGen, bmg *BishopMoveGen) *QueenMoveGen {
	return &QueenMoveGen{
		rookMoveGen:   rmg,
		bishopMoveGen: bmg,
	}
}

func (qmg *QueenMoveGen) GeneratePseudoLegalMoves(pos *position.Position) []Move {
	isWhite := pos.ColorToMove == constants.ColorWhite

	var otherColor uint8
	var piece uint8
	if isWhite {
		otherColor = constants.ColorBlack
		piece = constants.WhiteQueen
	} else {
		otherColor = constants.ColorWhite
		piece = constants.BlackQueen
	}

	moves := []Move{}
	bb := pos.PieceBitboards[piece]

	for bb != 0 {
		fromSquare := uint8(bits.TrailingZeros64(bb))

		attacks := qmg.GetAttacks(fromSquare, pos.ColorBitboards[constants.ColorBoth]) & ^pos.ColorBitboards[pos.ColorToMove]

		for attacks != 0 {
			toSquare := uint8(bits.TrailingZeros64(attacks))

			flags := FlagNone
			if bitboard.GetBit(pos.ColorBitboards[otherColor], toSquare) {
				flags = FlagCapture
			}

			moves = append(moves, Move{
				FromSquare: fromSquare,
				ToSquare:   toSquare,
				Piece:      piece,
				Flags:      flags,
			})

			bitboard.ClearBit(&attacks, toSquare)
		}

		bitboard.ClearBit(&bb, fromSquare)
	}

	return moves
}
